### PR TITLE
SONAR-8986 Add a method to retrieve the worker UUIDs

### PR DESCRIPTION
**Pull request by ehartmann**

### CODE REVIEW

File: server/sonar-ce/src/main/java/org/sonar/ce/container/ComputeEngineContainerImpl.java [hunk 1]
```diff
@@ -173,6 +175,14 @@ public ComputeEngineContainer start(Props props) {
 
     this.level4 = level3.createChild();
     this.level4.add(level4Components());
+
+    // TODO refactoring levelXComponents()
```
Comment (SonarTech) [inline server/sonar-ce/src/main/java/org/sonar/ce/container/ComputeEngineContainerImpl.java]: INFO Complete the task associated to this TODO comment. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1135)

**Comment by SonarTech** [comment]
SonarQube analysis reported 3 issues
* ![MAJOR][MAJOR] 1 major
* ![INFO][INFO] 2 info

Watch the comments in this conversation to review them.

#### 2 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [CeWorkerImpl.java#L97](https://github.com/SonarSource/sonarqube/blob/be956a7ab832a55b5bcd9cf3f8ff74c46a43c703/server/sonar-server/src/main/java/org/sonar/server/computation/taskprocessor/CeWorkerImpl.java#L97): Catch Exception instead of Throwable. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1181)
1. ![INFO][INFO] [CeWorkerImpl.java#L88](https://github.com/SonarSource/sonarqube/blob/be956a7ab832a55b5bcd9cf3f8ff74c46a43c703/server/sonar-server/src/main/java/org/sonar/server/computation/taskprocessor/CeWorkerImpl.java#L88): Complete the task associated to this TODO comment. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1135)

[INFO]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png 'Severity: INFO'
[MAJOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR'

File: server/sonar-ce/src/main/java/org/sonar/ce/container/ComputeEngineContainerImpl.java (hunk 1 above)
Comment (ehartmann) [inline server/sonar-ce/src/main/java/org/sonar/ce/container/ComputeEngineContainerImpl.java]: This will be done in another ticket.

File: server/sonar-server/src/main/java/org/sonar/server/computation/taskprocessor/CeWorkerFactory.java [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+package org.sonar.server.computation.taskprocessor;
+
+import java.util.List;
+
+public interface CeWorkerFactory {
```
Comment (sns-seb) [inline server/sonar-server/src/main/java/org/sonar/server/computation/taskprocessor/CeWorkerFactory.java]: please add documentation of your interface and its methods.

I find it inconsistent to have both a Listener pattern in place and a method getWorkerUUIDs(). Seems to me the pattern is used half way. Method `Listener.onChange` should take a set of workerUuids.

File: server/sonar-server/src/main/java/org/sonar/server/computation/taskprocessor/CeWorkerFactory.java [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+  CeWorker create();
+  List<String> getWorkerUUIDs();
```
Comment (sns-seb) [inline server/sonar-server/src/main/java/org/sonar/server/computation/taskprocessor/CeWorkerFactory.java]: should be a set, order doesn't matter but uniqueness does

File: server/sonar-ce/src/main/java/org/sonar/ce/cluster/ClusterClient.java [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+
+    // Create the Hazelcast instance
+    this.hzInstance = HazelcastClient.newHazelcastClient(hzConfig);
+    this.workerUUIDS = hzInstance.getReplicatedMap(WORKER_UUIDS);
+    this.ceWorkerFactory = ceWorkerFactory;
+    ceWorkerFactory.addListener(this);
+    this.workerUUIDS.put(hzInstance.getLocalEndpoint().getUuid(), ceWorkerFactory.getWorkerUUIDs());
+  }
+
+  public List<String> getWorkerUUIDs() {
```
Comment (sns-seb) [inline server/sonar-ce/src/main/java/org/sonar/ce/cluster/ClusterClient.java]: should be a set, order doesn't matter but uniqueness does

File: server/sonar-ce/src/main/java/org/sonar/ce/cluster/ClusterClient.java [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+    return workerUUIDS.values().stream().flatMap(List::stream).collect(Collectors.toList());
```
Comment (sns-seb) [inline server/sonar-ce/src/main/java/org/sonar/ce/cluster/ClusterClient.java]: this implies unit testing is lacking, badly, as this is the main usecase

File: server/sonar-ce/src/main/java/org/sonar/ce/cluster/ClusterClient.java [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+  }
+
+  @Override
+  public void close() {
```
Comment (sns-seb) [inline server/sonar-ce/src/main/java/org/sonar/ce/cluster/ClusterClient.java]: implementation of `AutoCloseable` is useless.
This method won't be called when CE shuts down.
If your intent was to have this method called when CE stops properly, then you should implement `org.picocontainer.Startable`

File: server/sonar-server/src/main/java/org/sonar/server/computation/taskprocessor/CeWorkerFactory.java (hunk 2 above)
Comment (sns-seb) [inline server/sonar-server/src/main/java/org/sonar/server/computation/taskprocessor/CeWorkerFactory.java]: actually, the use of the Listener pattern is useless.

It's point seems to be to workaround the constraint of order of instantiation between `ClusterClient`, `CeWorkerFactory` and `CeProcessingSchedulerImpl` (where `CeWorkerFactory` is called) but it totally fails at that. 

The dependency still exists:
if `ClusterClient` is instantiated after `CeProcessingSchedulerImpl`, the listener will be added after all the calls to `CeWorkerFactory#create` which makes the listener useless.
It seems you are aware of this limitation as you added the following to the constructor of `ClusterClient`:
```
this.workerUUIDS.put(hzInstance.getLocalEndpoint().getUuid(), ceWorkerFactory.getWorkerUUIDs());
```

As I pointed during our discussion last week, it will be way more reliable and explicite to actually code this dependency in SQ's code:

1. add method `initialize` to `ClusterClient` (or whatever more appropriate name) which implementation is to call on WorkerFactory to get the current set of worker uuids and share it across the cluster
2. add a call to this method after this line https://github.com/SonarSource/sonarqube/blob/master/server/sonar-server/src/main/java/org/sonar/server/computation/queue/CeQueueInitializer.java#L59

This approach has the benefit of not sharing the uuid of workers before they are actually started. A failure can occur between the time `CeProcessingSchedulerImpl` is instantiated and `CeQueueInitializer` runs.

File: server/sonar-server/src/main/java/org/sonar/server/computation/taskprocessor/CeWorker.java [hunk 7]
```diff
@@ -28,5 +28,5 @@
  * {@link Callable#call()} returns a Boolean which is {@code true} when some a {@link CeTask} was processed,
  * {@code false} otherwise.
  */
-public interface CeWorkerCallable extends Callable<Boolean> {
+public interface CeWorker extends Callable<Boolean> {
```
Comment (sns-seb) [inline server/sonar-server/src/main/java/org/sonar/server/computation/taskprocessor/CeWorker.java]: this renaming does not bring any value to the commit
if really necessary, it should be a dedicated commit

File: server/sonar-server/src/main/java/org/sonar/server/computation/taskprocessor/CeProcessingSchedulerImpl.java [hunk 8]
```diff
@@ -39,16 +39,17 @@
...
   private final TimeUnit timeUnit;
   private final ChainingCallback[] chainingCallbacks;
 
   public CeProcessingSchedulerImpl(CeConfiguration ceConfiguration,
-    CeProcessingSchedulerExecutorService processingExecutorService, CeWorkerCallable workerRunnable) {
+                                   CeProcessingSchedulerExecutorService processingExecutorService,
+                                   CeWorkerFactory ceWorkerFactory) {
     this.executorService = processingExecutorService;
-    this.workerRunnable = workerRunnable;
+    this.workerRunnable = ceWorkerFactory.create();
```
Comment (sns-seb) [inline server/sonar-server/src/main/java/org/sonar/server/computation/taskprocessor/CeProcessingSchedulerImpl.java]: this implies that unit tests are missing

File: server/sonar-ce/src/test/java/org/sonar/ce/container/ComputeEngineContainerImplTest.java [hunk 9]
```diff
@@ -63,19 +67,63 @@ public void constructor_does_not_create_container() {
...
+    properties.setProperty(ProcessProperties.CLUSTER_NAME, CLUSTER_NAME);
+
+    // required persisted properties
+    insertProperty(CoreProperties.SERVER_ID, "a_startup_id");
+    insertProperty(CoreProperties.SERVER_STARTTIME, DateUtils.formatDateTime(new Date()));
+
+    underTest
+      .start(new Props(properties));
+
+    MutablePicoContainer picoContainer = underTest.getComponentContainer().getPicoContainer();
```
Comment (sns-seb) [inline server/sonar-ce/src/test/java/org/sonar/ce/container/ComputeEngineContainerImplTest.java]: don't fork all this assertions on pico containers, it's going to be a pain to maintain.

Either make you test more specific (I guess you wanted to check classes `ClusterClient` and `ClusterProperties` are added, so just test that and rename the class to describe exactly that) or factor the testing code.

I favor the first solution which you can complete by added an assertion to `real_start_without_cluster` which ensure the classes aren't there.

File: server/sonar-ce/src/main/java/org/sonar/ce/cluster/ClusterClient.java [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+      .setProperty("hazelcast.tcp.join.port.try.count", "10")
+      // Don't bind on all interfaces
+      .setProperty("hazelcast.socket.bind.any", "false")
+      // Don't phone home
+      .setProperty("hazelcast.phone.home.enabled", "false")
+      // Use slf4j for logging
+      .setProperty("hazelcast.logging.type", "slf4j");
+
+    // Create the Hazelcast instance
+    this.hzInstance = HazelcastClient.newHazelcastClient(hzConfig);
```
Comment (sns-seb) [inline server/sonar-ce/src/main/java/org/sonar/ce/cluster/ClusterClient.java]: constructor should not contain long operations and avoid IO related operations.
Use `org.picocontainer.Startable#start` instead

File: server/sonar-server/src/main/java/org/sonar/server/computation/taskprocessor/CeProcessingSchedulerImpl.java (hunk 8 above)
Comment (sns-seb) [inline server/sonar-server/src/main/java/org/sonar/server/computation/taskprocessor/CeProcessingSchedulerImpl.java]: (adding my comment edit here to trigger notification)

the factory should be called at this line https://github.com/SonarSource/sonarqube/blob/master/server/sonar-server/src/main/java/org/sonar/server/computation/taskprocessor/CeProcessingSchedulerImpl.java#L59 as parameter of the `ChainingCallback` class which should now hold a `CeWorker` property rather than use the one of the `CeProcessingSchedulerImpl`

File: server/sonar-ce/src/main/java/org/sonar/ce/cluster/ClusterClient.java [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+import com.hazelcast.core.HazelcastInstance;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+import org.sonar.server.computation.taskprocessor.CeWorkerFactory;
+
+import static com.google.common.base.Preconditions.checkState;
+import static org.apache.commons.lang.StringUtils.isNotEmpty;
+
+public class ClusterClient implements AutoCloseable, CeWorkerFactory.Listener {
```
Comment (ehartmann) [inline server/sonar-ce/src/main/java/org/sonar/ce/cluster/ClusterClient.java]: Ok, I'll add this. In a cluster the WorkerFactory holds the UUIDs list

File: server/sonar-ce/src/main/java/org/sonar/ce/cluster/ClusterClientProperties.java [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+
+package org.sonar.ce.cluster;
+
+import org.sonar.api.config.Settings;
+import org.sonar.process.ProcessProperties;
+
+/**
+ * Properties of the cluster configuration
+ */
+public final class ClusterClientProperties {
```
Comment (ehartmann) [inline server/sonar-ce/src/main/java/org/sonar/ce/cluster/ClusterClientProperties.java]: It's just useful instead of looking at different properties and isolate the properties used.
But no problem it can be deleted.

File: server/sonar-ce/src/test/java/org/sonar/ce/cluster/TestSettings.java [hunk 13]
```diff
@@ -0,0 +1,68 @@
...
+
+import java.util.Map;
+import java.util.Optional;
+import java.util.Properties;
+import java.util.stream.Collectors;
+import org.sonar.api.config.Encryption;
+import org.sonar.api.config.PropertyDefinitions;
+import org.sonar.api.config.Settings;
+
+public class TestSettings extends Settings {
```
Comment (ehartmann) [inline server/sonar-ce/src/test/java/org/sonar/ce/cluster/TestSettings.java]: It was used in other place of testing.

**Comment by ehartmann** [comment]
This PR won't not be merged.
